### DISCUSSION TITLE
libc interop for opaque types

### DISCUSSION THREAD

**Discussion by jedbrown** [rust-lang/rust-bindgen#2599] [Q&A] [answered]
I'm using bindgen to process a header for a library that accepts arguments of type `FILE*`, which I've been handling using `Builder::opaque_type`. I'd like other parts of the code to be able to use [`libc::FILE`](https://docs.rs/libc/latest/libc/enum.FILE.html) in a way that matches the definition used by bindgen, but I can't find a way to do this. Is it possible (or a feature that makes sense to add) to tell bindgen that a given type is provided by the likes of `libc::FILE`?

As a concrete example, I'd like to be able to release a `libc-memstream` crate that provides a safe interface to [`libc::open_memstream`](https://docs.rs/libc/latest/libc/fn.open_memstream.html) to improve reliability when high-level bindings use it to extract strings from raw FFI interfaces that write information to a `FILE*` (like `fputs` and `fprintf`).

**Comment by pvdrz** [comment, marked as answer]
There's no feature that does exactly that, but you can do something like it by combining two features:

- First you need to call `blocklist_type` with `FILE` so there's no definition for `FILE` anywhere in your bindings. Given that `blocklist` methods are not recursive/transitive, all the uses of `FILE` in your bindings will still be present.

- Then you call `raw_line` with `use libc::FILE;`. Given that `libc::FILE` and `FILE` are the same type, they should be ABI compatible.

Let me know if this fixes your issue.

**Reply by jedbrown** [reply]
Thank you! This does what I need:
```rs
        .opaque_type("FILE") // don't chase the underlying types
        .blocklist_type("FILE") // and don't write a type in generated bindings
        .raw_line("use libc::FILE;") // write this instead
```

(If I leave out the first line, I get lots of warnings about the likes of `_IO_read_ptr` not using snake case.)

**Reply by pvdrz** [reply]
Oh yeah I didn't include it because you were already marking it as opaque :sweat_smile: but yeah, you also need that :)
